### ISSUE TITLE
Minimize ZIO Test type aliases 

### ISSUE THREAD

**Issue by jdegoes** [zio/zio#3499]
For fewer levels of indirection.

**Comment by adamgfraser** [comment]
Good idea. Here is a list of type aliases in ZIO Test (excluding basic definitions of services and type aliases defined in mocking and laws packages) along with some comments.

```scala
type AssertResultM = BoolAlgebraM[Any, Nothing, AssertionValue]
type AssertResult = BoolAlgebra[AssertionValue]
type TestResult = BoolAlgebra[FailureDetails]
```

Don't think we can eliminate this without significant refactoring since we use `BoolAlgebra` parameterized on two different types significantly within the test framework.

```scala
type TestAspectAtLeastR[R] = TestAspect[Nothing, R, Nothing, Any]
type TestAspectPoly = TestAspect[Nothing, Any, Nothing, Any]
type PerTest.AtLeastR[R] = TestAspect.PerTest[Nothing, R, Nothing, Any]
type PerTest.Poly = TestAspect.PerTest[Nothing, Any, Nothing, Any]
```

These seem like pretty easy eliminations. They make things slightly more ergonomic writing test aspects but probably do make it harder to understand if you aren't a power user where these type aliases are worth learning.

```scala
type TestReport[-E] = (Duration, ExecutedSpec[E]) => URIO[TestLogger, Unit]
```

Could be made into a trait to be slightly more accessible.

```scala
type ZTestEnv = TestClock with TestConsole with TestRandom with TestSystem
```

Slightly more ergonomic way of writing signature for combinators that restore test environment but not used outside of that. Could be eliminated.

```scala
type ZTest[-R, +E] = ZIO[R, TestFailure[E], TestSuccess]
```

Infrequently used, could be eliminated.

```scala
type ZSpec[-R, +E] = Spec[R, TestFailure[E], TestSuccess]
```

Used a lot. Probably need to keep.

```scala
type ExecutedResult[+E] = Either[TestFailure[E], TestSuccess]
```

Could probably delete

```scala
type ExecutedSpec[+E] = Spec[Any, Nothing, ExecutedResult[E]]
```

Probably need to keep. Could consider making into a separate data type. Advantage would be executed specs could be pure which would clean up test reporter significantly. Disadvantage would be code duplication between two different spec data types.

```scala
type Annotated[+A] = (A, TestAnnotationMap)
```

Could delete if we wanted.

@jdegoes Let me know if you agree and I can go through and do this.

**Comment by jdegoes** [comment]
@adamgfraser I suppose we should also decide between:

1. Keeping the test alias (a good example is: `ZEnv`, yes it adds indirection but saves a lot of repetition and pain; I think it's worth it)
2. Inlining the type alias (maybe useful in cases where it's not pulling its weight and infrequently used)
3. Introducing a new type (useful when it represents a different domain with different common operators, etc.; can hide complexity of full type and lead to a better experience)

**Comment by adamgfraser** [comment]
@jdegoes Yes. So here would be my proposal:

**Keep**

```scala
// These three we need to keep
type AssertResultM = BoolAlgebraM[Any, Nothing, AssertionValue]
type AssertResult = BoolAlgebra[AssertionValue]
type TestResult = BoolAlgebra[FailureDetails]
// These two I could go either way on
type Annotated[+A] = (A, TestAnnotationMap)
type ZTestEnv = TestClock with TestConsole with TestRandom with TestSystem
```

**Inline**

```scala
// Okay with keeping these though eliminating them does help us achieve goal of this ticket
type TestAspectAtLeastR[R] = TestAspect[Nothing, R, Nothing, Any]
type TestAspectPoly = TestAspect[Nothing, Any, Nothing, Any]
type PerTest.AtLeastR[R] = TestAspect.PerTest[Nothing, R, Nothing, Any]
type PerTest.Poly = TestAspect.PerTest[Nothing, Any, Nothing, Any]
type ZTest[-R, +E] = ZIO[R, TestFailure[E], TestSuccess]
```

**Introduce New Type**

```scala
// This one is a fairly minor change
type TestReporter[-E] = (Duration, ExecutedSpec[E]) => URIO[TestLogger, Unit]
// This one is more major and we should think about
type ZSpec[-R, +E] = Spec[R, TestFailure[E], TestSuccess]
type ExecutedSpec[+E] = Spec[Any, Nothing, ExecutedResult[E]]
```

Most of these are pretty straightforward and we can just agree and make the change.

The one I think we should consider more is splitting out `Spec` into `ZSpec` and `ExecutedSpec`. In the original version the current approach worked very well because by using the fixed point encoding we could use `Spec` to represent both specs where the values were ZIO effects as well as specs where the values were just results, so we could unify both effectual and executed tests. However, with the change to make suites effectual and now managed this has broken down so now we have executed specs being managed effects which isn't really what we want.

We could just have two different data types for `ZSpec` and `ExecutedSpec`. This would create some code duplication but it could clean up the `DefaultTestReporter` a lot and to a certain extent it would allow us to have methods that make sense for each type of spec (e.g. you would never provide an environment to an executed spec, and you would almost never want to use `exists` or `forall` on a `ZSpec` because just evaluating the predicate would require acquiring any resources the suite needs. If we did this we might also want to revisit the fixed point encoding as I think it makes the code somewhat more complex and I'm not sure we're getting a return on that investment today.

What do you think?

**Comment by jdegoes** [comment]
How about `case class Annotated[+A](value: A, annotationMap: TestAnnotationMap)`?

Agreed on the other ones and on `ExecutedSpec`, which I might rename to `SpecResults` in the refactoring.

For the test aspect ones: we already decided users won't often create these, I'm ok leaving them in for our convenience if you are ok with that.

**Comment by adamgfraser** [comment]
Sounds good!
